src/validationinterface.h

```diff
@@ -132,14 +132,19 @@ class CValidationInterface {
     virtual void TransactionRemovedFromMempool(const CTransactionRef &ptx) {}
     /**
      * Notifies listeners of a block being connected.
-     * Provides a vector of transactions evicted from the mempool as a result.
+     *
+     * The mempool is not guaranteed to be in a consistent state when this
+     * notification is fired. Listeners should use UpdatedBlockTip for that.
```

Comment:
Beginner question: what do you mean by "consistent state" here?

i.e. what could go wrong with the mempool after some, but not all, of the `ActivateBestChainStep` involved in `ActivateBestChain`?

Reply:
cs_main and mempool.cs are both still being held when the `BlockConnected` signal is called, and `UpdateMempoolForReorg()` may have been called after blocks were connected but before the `BlockConnected` signal is called. See the comment above for the `TransactionAddedToMempool` signal.

The `UpdatedBlockTip` signal is called just before releasing `cs_main`, when we know that the mempool is consistent with the chain tip.

Reply:
I agree with your first point. For the second point, if we consider the client to be a passive listener (ie the only input it's getting is from the validation interface), then it shouldn't/can't acquire those locks. It should wait for a `UpdatedBlockTip` to be sure that the state is consistent.